receiver/prometheusexec/subprocessmanager/manager.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package subprocessmanager
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"math"
+	"math/rand"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/kballard/go-shellquote"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusexec/subprocessmanager/config"
+)
+
+// Process struct holds all the info needed to instantiate a subprocess
+type Process struct {
+	Command    string
+	Port       int
+	Env        []config.EnvConfig
+	CustomName string
+}
+
+const (
+	// HealthyProcessTime is the default time a process needs to stay alive to be considered healthy
+	HealthyProcessTime time.Duration = 30 * time.Minute
+	// HealthyCrashCount is the amount of times a process can crash (within the healthyProcessTime) before being considered unstable - it may be trying to find a port
+	HealthyCrashCount int = 3
+	// delayMutiplier is the factor by which the delay scales
+	delayMultiplier float64 = 2.0
+	// initialDelay is the initial delay before a process is restarted
+	initialDelay time.Duration = 1 * time.Second
+)
+
+// Run will start the process and keep track of running time
+func (proc *Process) Run(logger *zap.Logger) (time.Duration, error) {
+	var (
+		start     time.Time
+		elapsed   time.Duration
+		argsSlice []string
+	)
+
+	// Parse the command line string into arguments
+	args, err := shellquote.Split(proc.Command)
+	if err != nil {
+		return elapsed, fmt.Errorf("[%v] could not parse command error: %v", proc.CustomName, err)
```

Comment:
- For the error use `%w` instead of `%v`, see https://blog.golang.org/go1.13-errors
- For CustomName, this is not needed as `config.Name()` will already be included in the logs as dicsussed here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/370#discussion_r449736400. You mentioned a concern about this not applies to logs from the subprocesses, but that won't be an issue since you are using the same `logger` instance.

Same applies for most of this file

Reply:
You're right, thanks! Fixing